packages/services/src/kernel/future.ts

```diff
@@ -189,26 +190,20 @@ export class KernelFutureHandler<
     super.dispose();
   }
 
-  /**
-   * Handle an incoming kernel message.
-   */
-  async handleMsg(msg: KernelMessage.IMessage): Promise<void> {
+  protected async handleDefaultMsg(msg: KernelMessage.IMessage): Promise<void> {
```

Comment:
Ah, I was thinking we'd define a base `handleMsg` with this logic, then the subclass `handleMsg` would have similar switch statement with `super.handleMsg` called in the `default:` case. But this is getting into C++ish patterns, so I also defer to your experience of how this pattern might play out in C++.

Reply:
Something like:

```patch
diff --git a/packages/services/src/kernel/future.ts b/packages/services/src/kernel/future.ts
index fe354ae06..588432692 100644
--- a/packages/services/src/kernel/future.ts
+++ b/packages/services/src/kernel/future.ts
@@ -23,7 +23,6 @@ export abstract class KernelFutureHandler<
   REQUEST extends KernelMessage.IShellControlMessage,
   REPLY extends KernelMessage.IShellControlMessage
 > extends DisposableDelegate implements Kernel.IFuture<REQUEST, REPLY> {
-  abstract async handleMsg(msg: KernelMessage.IMessage): Promise<void>;
   /**
    * Construct a new KernelFutureHandler.
    */
@@ -190,7 +189,7 @@ export abstract class KernelFutureHandler<
     super.dispose();
   }
 
-  protected async handleDefaultMsg(msg: KernelMessage.IMessage): Promise<void> {
+  protected async handleMsg(msg: KernelMessage.IMessage): Promise<void> {
     switch (msg.channel) {
       case 'stdin':
         await this.handleStdin(msg as KernelMessage.IStdinMessage);
@@ -294,10 +293,13 @@ export class KernelControlFutureHandler<
    * Handle an incoming kernel message.
    */
   async handleMsg(msg: KernelMessage.IMessage): Promise<void> {
-    if (msg.channel === 'control') {
-      await this.handleReply(msg as REPLY);
-    } else {
-      await this.handleDefaultMsg(msg);
+    switch (msg.channel) {
+      case 'control':
+        await this.handleReply(msg as REPLY);
+        break;
+      default:
+        await super.handleMsg(msg);
+        break;
     }
   }
 }
@@ -311,10 +313,13 @@ export class KernelShellFutureHandler<
    * Handle an incoming kernel message.
    */
   async handleMsg(msg: KernelMessage.IMessage): Promise<void> {
-    if (msg.channel === 'shell') {
-      await this.handleReply(msg as REPLY);
-    } else {
-      await this.handleDefaultMsg(msg);
+    switch (msg.channel) {
+      case 'shell':
+        await this.handleReply(msg as REPLY);
+        break;
+      default:
+        await super.handleMsg(msg);
+        break;
     }
   }
 }
```

Reply:
Or what about this, which does even better checking of the parent header, and moves all of the logic into the super class?

```patch
diff --git a/packages/services/src/kernel/future.ts b/packages/services/src/kernel/future.ts
index fe354ae06..b2678cc6f 100644
--- a/packages/services/src/kernel/future.ts
+++ b/packages/services/src/kernel/future.ts
@@ -23,7 +23,6 @@ export abstract class KernelFutureHandler<
   REQUEST extends KernelMessage.IShellControlMessage,
   REPLY extends KernelMessage.IShellControlMessage
 > extends DisposableDelegate implements Kernel.IFuture<REQUEST, REPLY> {
-  abstract async handleMsg(msg: KernelMessage.IMessage): Promise<void>;
   /**
    * Construct a new KernelFutureHandler.
    */
@@ -190,8 +189,14 @@ export abstract class KernelFutureHandler<
     super.dispose();
   }
 
-  protected async handleDefaultMsg(msg: KernelMessage.IMessage): Promise<void> {
+  protected async handleMsg(msg: KernelMessage.IMessage): Promise<void> {
     switch (msg.channel) {
+      case 'control':
+      case 'shell':
+        if (msg.parent_header === this.msg.header) {
+          await this.handleReply(msg as REPLY);
+        }
+        break;
       case 'stdin':
         await this.handleStdin(msg as KernelMessage.IStdinMessage);
         break;
@@ -289,35 +294,13 @@ export class KernelControlFutureHandler<
   REQUEST extends KernelMessage.IControlMessage = KernelMessage.IControlMessage,
   REPLY extends KernelMessage.IControlMessage = KernelMessage.IControlMessage
 > extends KernelFutureHandler<REQUEST, REPLY>
-  implements Kernel.IControlFuture<REQUEST, REPLY> {
-  /**
-   * Handle an incoming kernel message.
-   */
-  async handleMsg(msg: KernelMessage.IMessage): Promise<void> {
-    if (msg.channel === 'control') {
-      await this.handleReply(msg as REPLY);
-    } else {
-      await this.handleDefaultMsg(msg);
-    }
-  }
-}
+  implements Kernel.IControlFuture<REQUEST, REPLY> {}
 
 export class KernelShellFutureHandler<
   REQUEST extends KernelMessage.IShellMessage = KernelMessage.IShellMessage,
   REPLY extends KernelMessage.IShellMessage = KernelMessage.IShellMessage
 > extends KernelFutureHandler<REQUEST, REPLY>
-  implements Kernel.IShellFuture<REQUEST, REPLY> {
-  /**
-   * Handle an incoming kernel message.
-   */
-  async handleMsg(msg: KernelMessage.IMessage): Promise<void> {
-    if (msg.channel === 'shell') {
-      await this.handleReply(msg as REPLY);
-    } else {
-      await this.handleDefaultMsg(msg);
-    }
-  }
-}
+  implements Kernel.IShellFuture<REQUEST, REPLY> {}
 
 namespace Private {
   /**
```

Reply:
I went ahead and pushed changes similar to these to this PR. One nice thing is that the previously private methods can still remain private. (of course, this can always be opened up again in a minor release, but it's nice that the api surface isn't expanding)